### PR TITLE
Pipeline, Alisa Karpova - 22FPL2

### PR DESCRIPTION
**Pull request by AlisaKarpova**

### CODE REVIEW

**Comment by marina-kaz** [comment]
No changes in pipeline to review today

File: lab_6_pipeline/pipeline.py [hunk 1]
```diff
@@ -27,16 +53,46 @@ def __init__(self, path_to_raw_txt_data: pathlib.Path) -> None:
...
+        if len(raw_files) != len(meta_files):
+            raise InconsistentDatasetError
+
+        sorted_raw_files = sorted(raw_files, key=lambda x: get_article_id_from_filepath(x))
+        sorted_meta_files = sorted(meta_files, key=lambda x: get_article_id_from_filepath(x))
+
+        for index, (raw_file, meta_file) in enumerate(zip(sorted_raw_files, sorted_meta_files)):
+            if (index + 1 != get_article_id_from_filepath(raw_file)
+                    or index + 1 != get_article_id_from_filepath(meta_file)
+                    or raw_file.stat().st_size == 0 or meta_file.stat().st_size == 0):
```
Comment (marina-kaz) [inline lab_6_pipeline/pipeline.py:85]: `== 0` is redundant. Can you explain why?

File: lab_6_pipeline/pipeline.py [hunk 2]
```diff
@@ -27,16 +53,46 @@ def __init__(self, path_to_raw_txt_data: pathlib.Path) -> None:
...
+        raw_files = list(self.path_to_raw_txt_data.glob('*_raw.txt'))
+        meta_files = list(self.path_to_raw_txt_data.glob('*_meta.json'))
+        if len(raw_files) != len(meta_files):
+            raise InconsistentDatasetError
+
+        sorted_raw_files = sorted(raw_files, key=lambda x: get_article_id_from_filepath(x))
+        sorted_meta_files = sorted(meta_files, key=lambda x: get_article_id_from_filepath(x))
+
+        for index, (raw_file, meta_file) in enumerate(zip(sorted_raw_files, sorted_meta_files)):
+            if (index + 1 != get_article_id_from_filepath(raw_file)
```
Comment (marina-kaz) [inline lab_6_pipeline/pipeline.py:83]: Can we avoid adding 1 at each step by directly setting `enumerate` to start from 1?

File: lab_6_pipeline/pipeline.py [hunk 3]
```diff
@@ -62,12 +119,21 @@ def __init__(
             corpus_manager (CorpusManager): CorpusManager instance
             analyzer (LibraryWrapper | None): Analyzer instance
         """
+        self._corpus_manager = corpus_manager
+        self.analyzer = analyzer
```
Comment (marina-kaz) [inline lab_6_pipeline/pipeline.py:123]: Why public?

File: lab_6_pipeline/pipeline.py [hunk 4]
```diff
@@ -62,12 +119,21 @@ def __init__(
...
     def run(self) -> None:
         """
         Perform basic preprocessing and write processed text to files.
         """
-
+        articles = self._corpus_manager.get_articles().values()
+        for article in articles:
+            to_cleaned(article)
+            if self.analyzer:
+                texts = split_by_sentence(article.text)
```
Comment (marina-kaz) [inline lab_6_pipeline/pipeline.py:133]: Why?

**Review by marina-kaz — COMMENTED** [review]
Good job!
Maybe try for mark 8?

File: lab_6_pipeline/pipeline.py (hunk 2 above)
Comment (AlisaKarpova) [reply lab_6_pipeline/pipeline.py:83]: Fixed

File: lab_6_pipeline/pipeline.py (hunk 3 above)
Comment (AlisaKarpova) [reply lab_6_pipeline/pipeline.py:123]: Changed it, it should be private

File: lab_6_pipeline/pipeline.py (hunk 4 above)
Comment (AlisaKarpova) [reply lab_6_pipeline/pipeline.py:133]: I tried to change it

File: lab_6_pipeline/pipeline.py [hunk 5]
```diff
@@ -189,6 +295,14 @@ def _count_frequencies(self, article: Article) -> dict[str, int]:
         Returns:
             dict[str, int]: POS frequencies
         """
+        frequencies = {}
+        for conllu_sentence in self._analyzer.from_conllu(article).sentences:
+            for word in conllu_sentence.words:
+                word_feature = word.to_dict()['upos']
+                if word_feature not in frequencies:
+                    frequencies[word_feature] = frequencies.get(word_feature, 0)
```
Comment (marina-kaz) [inline lab_6_pipeline/pipeline.py:303]: Why use `get` if you've established that the item is not the dictionary? 
You may skip the condition altogether.

Comment (AlisaKarpova) [reply lab_6_pipeline/pipeline.py:303]: Tried to change it

File: lab_6_pipeline/pipeline.py [hunk 6]
```diff
@@ -107,6 +181,10 @@ def to_conllu(self, article: Article) -> None:
         Args:
             article (Article): Article containing information to save
         """
+        with open(article.get_file_path(kind=ArtifactType.UDPIPE_CONLLU), 'w',
+                  encoding='utf-8') as annotation_file:
+            annotation_file.writelines(article.get_conllu_info())
```
[the remark this answers is not in the document]
Comment (AlisaKarpova) [reply lab_6_pipeline/pipeline.py:186]: Fixed

File: lab_6_pipeline/pipeline.py (hunk 1 above)
Comment (AlisaKarpova) [reply lab_6_pipeline/pipeline.py:85]: Changed it. It returns 0 if the file is empty

File: lab_6_pipeline/pipeline.py [hunk 7]
```diff
@@ -173,11 +267,23 @@ def __init__(self, corpus_manager: CorpusManager, analyzer: LibraryWrapper) -> N
...
         """
+        self._corpus_manager = corpus_manager
+        self._analyzer = analyzer
 
     def run(self) -> None:
         """
         Visualize the frequencies of each part of speech.
         """
+        for article_id, article in self._corpus_manager.get_articles().items():
+            if not article.get_file_path(kind=ArtifactType.STANZA_CONLLU).stat().st_size == 0:
```
[the remark this answers is not in the document]
Comment (AlisaKarpova) [reply lab_6_pipeline/pipeline.py:278]: Fixed

File: lab_6_pipeline/pipeline.py [hunk 8]
```diff
@@ -173,11 +267,23 @@ def __init__(self, corpus_manager: CorpusManager, analyzer: LibraryWrapper) -> N
...
+                raise EmptyFileError
+            from_meta(article.get_meta_file_path(), article)
+            article.set_pos_info(self._count_frequencies(article))
+            to_meta(article)
+            visualize(article=article,
+                      path_to_save=self._corpus_manager.path_to_raw_txt_data /
```
[the remark this answers is not in the document]
Comment (AlisaKarpova) [reply lab_6_pipeline/pipeline.py]: Fixed

**Pull request closed without merging**